File: Examples/005/Fort/Problem/aos_soa.f90
program aos_soa
    implicit none
    integer, parameter :: NUM_PARTICLES = 10000000
    integer, parameter :: ITERATIONS = 200
    integer :: i, iter
    real(8) :: dt = 0.01d0

    ! BAD: Array of Structures (AoS) via Derived Types
    type :: particle
        real(8) :: x, y, z
        real(8) :: vx, vy, vz
    end type particle
    type(particle), allocatable :: aos(:)

    ! GOOD: Structure of Arrays (SoA)
    real(8), allocatable :: soa_x(:), soa_y(:), soa_z(:)
    real(8), allocatable :: soa_vx(:), soa_vy(:), soa_vz(:)

    ! Initialization
    allocate(aos(NUM_PARTICLES))
    allocate(soa_x(NUM_PARTICLES), soa_y(NUM_PARTICLES), soa_z(NUM_PARTICLES))
    allocate(soa_vx(NUM_PARTICLES), soa_vy(NUM_PARTICLES), soa_vz(NUM_PARTICLES))
    
    do i = 1, NUM_PARTICLES
        aos(i)%x = 1.0d0; aos(i)%y = 1.0d0; aos(i)%z = 1.0d0
        aos(i)%vx = 0.1d0; aos(i)%vy = 0.1d0; aos(i)%vz = 0.1d0
    end do
    soa_x = 1.0d0; soa_y = 1.0d0; soa_z = 1.0d0
    soa_vx = 0.1d0; soa_vy = 0.1d0; soa_vz = 0.1d0

    ! --- AoS Update ---
    do iter = 1, ITERATIONS
        do i = 1, NUM_PARTICLES
            aos(i)%x = aos(i)%x + aos(i)%vx * dt
            aos(i)%y = aos(i)%y + aos(i)%vy * dt
            aos(i)%z = aos(i)%z + aos(i)%vz * dt
        end do
    end do

    ! --- SoA Update ---
    do iter = 1, ITERATIONS
        do i = 1, NUM_PARTICLES
            soa_x(i) = soa_x(i) + soa_vx(i) * dt
            soa_y(i) = soa_y(i) + soa_vy(i) * dt
            soa_z(i) = soa_z(i) + soa_vz(i) * dt
        end do
    end do

    print *, "Done. Sample AoS X: ", aos(1)%x, ", SoA X: ", soa_x(1)
end program aos_soa

File: Examples/005/CPP/Problem/aos_soa.cpp
#include <iostream>
#include <vector>

const int NUM_PARTICLES = 10000000;
const int ITERATIONS = 200;

// BAD: Array of Structures (AoS)
struct Particle {
    double x, y, z;
    double vx, vy, vz;
};

// GOOD: Structure of Arrays (SoA)
struct ParticlesSoA {
    std::vector<double> x, y, z;
    std::vector<double> vx, vy, vz;
    
    ParticlesSoA(int size) {
        x.resize(size, 1.0); y.resize(size, 1.0); z.resize(size, 1.0);
        vx.resize(size, 0.1); vy.resize(size, 0.1); vz.resize(size, 0.1);
    }
};

void updateAoS(std::vector<Particle>& p, double dt) {
    for (int iter = 0; iter < ITERATIONS; ++iter) {
        for (int i = 0; i < NUM_PARTICLES; ++i) {
            p[i].x += p[i].vx * dt;
            p[i].y += p[i].vy * dt;
            p[i].z += p[i].vz * dt;
        }
    }
}

void updateSoA(ParticlesSoA& p, double dt) {
    for (int iter = 0; iter < ITERATIONS; ++iter) {
        // The compiler will easily vectorize these tight, contiguous loops
        for (int i = 0; i < NUM_PARTICLES; ++i) p.x[i] += p.vx[i] * dt;
        for (int i = 0; i < NUM_PARTICLES; ++i) p.y[i] += p.vy[i] * dt;
        for (int i = 0; i < NUM_PARTICLES; ++i) p.z[i] += p.vz[i] * dt;
    }
}

int main() {
    std::vector<Particle> aos(NUM_PARTICLES, {1.0, 1.0, 1.0, 0.1, 0.1, 0.1});
    ParticlesSoA soa(NUM_PARTICLES);
    
    double dt = 0.01;

    // We call both so you can compare them side-by-side in Score-P
    updateAoS(aos, dt);
    updateSoA(soa, dt);

    std::cout << "Done. Sample AoS X: " << aos[0].x << ", SoA X: " << soa.x[0] << std::endl;
    return 0;
}
